File: backend/src/main/java/com/urbanman/ecommerce/controller/PaymentController.java
package com.urbanman.ecommerce.controller;

import com.razorpay.Order;
import com.razorpay.RazorpayClient;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/payment")
public class PaymentController {

    @Value("${razorpay.key.id:}")
    private String keyId;

    @Value("${razorpay.key.secret:}")
    private String keySecret;

    // Create a Razorpay order
    @PostMapping("/create-order")
    public ResponseEntity<?> createOrder(@RequestBody Map<String, Object> body) {
        if (keyId == null || keyId.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Razorpay not configured"));
        }
        try {
            double amount = Double.parseDouble(body.get("amount").toString());
            RazorpayClient client = new RazorpayClient(keyId, keySecret);

            JSONObject opts = new JSONObject();
            opts.put("amount", (int)(amount * 100)); // paise
            opts.put("currency", "INR");
            opts.put("receipt", "order_" + System.currentTimeMillis());

            Order rzpOrder = client.orders.create(opts);

            Map<String, Object> result = new HashMap<>();
            result.put("orderId",  rzpOrder.get("id"));
            result.put("amount",   rzpOrder.get("amount"));
            result.put("currency", rzpOrder.get("currency"));
            result.put("keyId",    keyId);
            return ResponseEntity.ok(result);
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body(Map.of("error", e.getMessage()));
        }
    }

    // Verify payment signature
    @PostMapping("/verify")
    public ResponseEntity<?> verifyPayment(@RequestBody Map<String, String> body) {
        try {
            String razorpayOrderId   = body.get("razorpay_order_id");
            String razorpayPaymentId = body.get("razorpay_payment_id");
            String razorpaySignature = body.get("razorpay_signature");

            String data = razorpayOrderId + "|" + razorpayPaymentId;
            javax.crypto.Mac mac = javax.crypto.Mac.getInstance("HmacSHA256");
            mac.init(new javax.crypto.spec.SecretKeySpec(keySecret.getBytes(), "HmacSHA256"));
            byte[] hash = mac.doFinal(data.getBytes());
            StringBuilder hex = new StringBuilder();
            for (byte b : hash) hex.append(String.format("%02x", b));

            boolean valid = hex.toString().equals(razorpaySignature);
            return ResponseEntity.ok(Map.of("verified", valid));
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body(Map.of("error", e.getMessage()));
        }
    }
}

File: backend/controllers/paymentController.js
const Razorpay = require('razorpay');
const crypto = require('crypto');
require('dotenv').config();

const keyId = process.env.RAZORPAY_KEY_ID;
const keySecret = process.env.RAZORPAY_KEY_SECRET;

const createOrder = async (req, res) => {
  if (!keyId || !keySecret) {
    return res.status(400).json({ error: 'Razorpay not configured' });
  }

  try {
    const amount = parseFloat(req.body.amount);
    if (isNaN(amount) || amount <= 0) {
      return res.status(400).json({ error: 'Valid amount is required' });
    }

    const instance = new Razorpay({
      key_id: keyId,
      key_secret: keySecret
    });

    const options = {
      amount: Math.round(amount * 100), // Paise
      currency: 'INR',
      receipt: 'order_' + Date.now()
    };

    const rzpOrder = await instance.orders.create(options);

    return res.status(200).json({
      orderId: rzpOrder.id,
      amount: rzpOrder.amount,
      currency: rzpOrder.currency,
      keyId: keyId
    });
  } catch (error) {
    console.error('Create payment order error:', error);
    return res.status(500).json({ error: error.message || 'Internal server error' });
  }
};

const verifyPayment = async (req, res) => {
  if (!keySecret) {
    return res.status(400).json({ error: 'Razorpay not configured' });
  }

  try {
    const { razorpay_order_id, razorpay_payment_id, razorpay_signature } = req.body;

    if (!razorpay_order_id || !razorpay_payment_id || !razorpay_signature) {
      return res.status(400).json({ error: 'Missing signature parameters' });
    }

    const data = razorpay_order_id + '|' + razorpay_payment_id;
    const generated_signature = crypto
      .createHmac('sha256', keySecret)
      .update(data)
      .digest('hex');

    const valid = generated_signature === razorpay_signature;
    return res.status(200).json({ verified: valid });
  } catch (error) {
    console.error('Verify payment error:', error);
    return res.status(500).json({ error: error.message || 'Internal server error' });
  }
};

module.exports = {
  createOrder,
  verifyPayment
};
